### ISSUE TITLE
"use base" masks $SIG{__DIE__}

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8851]
Migrated from [rt.perl.org#42163](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=42163) (status was 'resolved')

Searchable as RT42163$

**Comment by p5pRT** [comment]
### From @schwern

#### Created by @schwern



If a module sets a global $SIG\{\_\_DIE\_\_\} handler and is loaded with "use base" that $SIG\{\_\_DIE\_\_\} handler will be localized and thus not set\.

This is because of this code in base\.pm\.

&nbsp;            local $SIG\{\_\_DIE\_\_\};
&nbsp;            eval "require $base";

aliased\.pm had a similar problem and I propose a similar fix\.
http&#8203;://rt\.cpan\.org/Public/Bug/Display\.html?id=25293

Patch forthcoming\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.8.8:

Configured by schwern at Sat Oct 14 18:47:46 PDT 2006.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=darwin, osvers=8.8.1, archname=darwin-thread-multi-2level
    uname='darwin windhund.local 8.8.1 darwin kernel version 8.8.1: mon sep 25 19:42:00 pdt 2006; root:xnu-792.13.8.obj~1release_i386 i386 i386 macbook1,1 darwin '
    config_args='-Dprefix=/usr/local/perl/5.8.8 -Duseithreads -Dccflags=-I/sw/include -Dldflags=-L/sw/lib -Dperladmin=schwern@pobox.com -Dcf_email=schwern@pobox.com -Dmyhostname=windhund -Dmydomain=.schwern.org -Dprivlib=/usr/local/perl/5.8.8/lib -Dsiteprefix=/usr/local -Dsitelib=/usr/local/lib/site_perl -Dlibpth=/usr/local/lib /sw/lib /opt/local/lib /usr/lib -Dnetdb_host_type=const char * -Dnetdb_hlen_type=socklen_t -Uinstallusrbinperl'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-I/sw/include -fno-common -DPERL_DARWIN -no-cpp-precomp -fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/opt/local/include',
    optimize='-O3',
    cppflags='-no-cpp-precomp -I/sw/include -fno-common -DPERL_DARWIN -no-cpp-precomp -fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/opt/local/include'
    ccversion='', gccversion='4.0.1 (Apple Computer, Inc. build 5363)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags ='-L/sw/lib -L/usr/local/lib -L/opt/local/lib'
    libpth=/usr/local/lib /sw/lib /opt/local/lib /usr/lib
    libs=-lgdbm -ldbm -ldl -lm -lc
    perllibs=-ldl -lm -lc
    libc=/usr/lib/libc.dylib, so=dylib, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-L/sw/lib -bundle -undefined dynamic_lookup -L/usr/local/lib -L/opt/local/lib'

Locally applied patches:
    


@INC for perl v5.8.8:
    /sw/lib/perl5
    /sw/lib/perl5/darwin
    /usr/local/perl/5.8.8/lib/darwin-thread-multi-2level
    /usr/local/perl/5.8.8/lib
    /usr/local/lib/site_perl/5.8/darwin-thread-multi-2level
    /usr/local/lib/site_perl
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.8:
    DYLD_LIBRARY_PATH (unset)
    HOME=/Users/schwern
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=en_US.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/opt/local/bin:/opt/local/sbin:/sw/bin:/sw/sbin:/Users/schwern/bin:/usr/local/bin:/Users/schwern/bin:/usr/local/bin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/X11R6/bin
    PERL5LIB=/sw/lib/perl5:/sw/lib/perl5/darwin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @schwern

Attached is the patch\.

**Comment by p5pRT** [comment]
### From @schwern

<details><summary>base_sigdie.patch</summary>

```diff
--- lib/base.pm	2006-10-23 01:52:58.000000000 -0700
+++ lib/base.pm	2007-03-28 21:42:53.000000000 -0700
@@ -82,21 +82,30 @@
 	      unless defined ${$base.'::VERSION'};
         }
         else {
-            local $SIG{__DIE__};
-            eval "require $base";
-            # Only ignore "Can't locate" errors from our eval require.
-            # Other fatal errors (syntax etc) must be reported.
-            die if $@ && $@ !~ /^Can't locate .*? at \(eval /;
-            unless (%{"$base\::"}) {
-                require Carp;
-                Carp::croak(<<ERROR);
+            my $sigdie;
+            {
+                local $SIG{__DIE__};
+                eval "require $base";
+                $sigdie = $SIG{__DIE__};
+
+                # Only ignore "Can't locate" errors from our eval require.
+                # Other fatal errors (syntax etc) must be reported.
+                die if $@ && $@ !~ /^Can't locate .*? at \(eval /;
+                unless (%{"$base\::"}) {
+                    require Carp;
+                    Carp::croak(<<ERROR);
 Base class package "$base" is empty.
     (Perhaps you need to 'use' the module which defines that package first.)
 ERROR
 
+                }
             }
+
             ${$base.'::VERSION'} = "-1, set by base.pm"
               unless defined ${$base.'::VERSION'};
+            
+            # Make sure a global $SIG{__DIE__} makes it out of the localization
+            $SIG{__DIE__} = $sigdie if defined $sigdie;
         }
         push @{"$inheritor\::ISA"}, $base;
 
--- lib/base/t/base.t	2006-10-23 01:52:58.000000000 -0700
+++ lib/base/t/base.t	2007-03-28 21:42:47.000000000 -0700
@@ -1,7 +1,8 @@
 #!/usr/bin/perl -w
 
 use strict;
-use Test::More tests => 11;
+use Test::More tests => 13;
+use lib 't/lib';
 
 use_ok('base');
 
@@ -85,3 +86,21 @@
          'Base class empty error message');
 
 }
+
+
+# Test that a global sigdie is preserved.
+{
+    local $SIG{__DIE__};
+    {
+        package Test::HasSigDie;
+        eval q[use base 'HasSigDie'];
+    }
+
+    ::ok $SIG{__DIE__}, 'sigdie set';
+    eval {
+        die "Testing sigdie handler";
+    };
+    package HasSigDie;
+    our $Sigdie_ran;
+    ::is $Sigdie_ran, 1;
+}
--- /dev/null	2007-03-28 21:42:35.000000000 -0700
+++ t/lib/HasSigDie.pm	2007-03-28 21:01:09.000000000 -0700
@@ -0,0 +1,5 @@
+package HasSigDie;
+
+$SIG{__DIE__} = sub { $Sigdie_ran++ };
+
+1;

```
</details>

**Comment by p5pRT** [comment]
@schwern - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @smpeters

On Wed\, Mar 28\, 2007 at 08&#8203;:50&#8203;:28PM \-0700\, Michael G Schwern wrote&#8203;:
> \# New Ticket Created by  Michael G Schwern 
> \# Please include the string&#8203;:  \[perl \#42163\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=42163 >
> 
> 
> This is a bug report for perl from schwern@&#8203;pobox\.com\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> If a module sets a global $SIG\{\_\_DIE\_\_\} handler and is loaded with "use base" that $SIG\{\_\_DIE\_\_\} handler will be localized and thus not set\.
> 
> This is because of this code in base\.pm\.
> 
>             local $SIG\{\_\_DIE\_\_\};
>             eval "require $base";
> 
> aliased\.pm had a similar problem and I propose a similar fix\.
> http&#8203;://rt\.cpan\.org/Public/Bug/Display\.html?id=25293
> 
> Patch forthcoming\.
> 

Since you are listed as the maintainer of base on CPAN\, would you be
interested in making a release?  If not\, would you be willing to work 
with someone to make a release?

Steve Peters
steve@&#8203;fisharerojo\.org

**Comment by p5pRT** [comment]
### From @schwern

Steve Peters via RT wrote&#8203;:
> Since you are listed as the maintainer of base on CPAN\, would you be
> interested in making a release?  If not\, would you be willing to work 
> with someone to make a release?

I'm not particularly interested in maintaining that module anymore and would
be happy to hand it off to somebody\.

**Comment by p5pRT** [comment]
### From @rgarcia

On 29/03/07\, Michael G Schwern \<schwern@&#8203;pobox\.com> wrote&#8203;:
> I'm not particularly interested in maintaining that module anymore and would
> be happy to hand it off to somebody\.

I could do it\, if you give me the rights on pause\.

**Comment by p5pRT** [comment]
### From @schwern

Rafael Garcia\-Suarez wrote&#8203;:
> On 29/03/07\, Michael G Schwern \<schwern@&#8203;pobox\.com> wrote&#8203;:
>> I'm not particularly interested in maintaining that module anymore and
>> would
>> be happy to hand it off to somebody\.
> 
> I could do it\, if you give me the rights on pause\.

Done\, but only for base\.  Ilya has primary maintenance of fields which ships
in the base distribution\.  Ilya?

**Comment by p5pRT** [comment]
### From nospam-abuse@ilyaz.org

On Fri\, Mar 30\, 2007 at 03&#8203;:26&#8203;:35AM \-0700\, Michael G Schwern wrote&#8203;:
> Rafael Garcia\-Suarez wrote&#8203;:
> > On 29/03/07\, Michael G Schwern \<schwern@&#8203;pobox\.com> wrote&#8203;:
> >> I'm not particularly interested in maintaining that module anymore and
> >> would
> >> be happy to hand it off to somebody\.
> > 
> > I could do it\, if you give me the rights on pause\.
> 
> Done\, but only for base\.  Ilya has primary maintenance of fields which ships
> in the base distribution\.  Ilya?

Looks like a misprint\.  I have never used fields\, so I doubt I
maintain it\.  ;\-\)

Yours\,
Ilya

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Fri\, 30 Mar 2007 04&#8203;:19&#8203;:58 \-0700\, Ilya Zakharevich \<nospam\-abuse@&#8203;ilyaz\.org> said&#8203;:

 >> Done\, but only for base\.  Ilya has primary maintenance of fields which ships
 >> in the base distribution\.  Ilya?

&nbsp;  > Looks like a misprint\.  I have never used fields\, so I doubt I
&nbsp;  > maintain it\.  ;\-\)

I seem to recall that this was a consequence of one of your OS2
packages for some reason\. Nonsense corrected\, now PAUSE

&nbsp;    "Made RGARCIA primary maintainer of fields\."

\-\- 
andreas

**Comment by p5pRT** [comment]
### From @rgarcia

On 30/03/07\, Andreas J\. Koenig \<andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de> wrote&#8203;:
> I seem to recall that this was a consequence of one of your OS2
> packages for some reason\. Nonsense corrected\, now PAUSE
>
>     "Made RGARCIA primary maintainer of fields\."

Thanks\. I'll upload a new version before the next stable perl :\)

**Comment by p5pRT** [comment]
### From @rgs

This problem has been fixed in bleadperl by change \#31163\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'
